src/Microsoft.Azure.SignalR.Common/ServiceConnections/scopes/ClientConnectionScope.cs

```diff
@@ -14,25 +15,37 @@ internal class ClientConnectionScope : IDisposable
     {
         private bool _needCleanup;
 
-        internal ClientConnectionScope() : this(default, default)
+        internal ClientConnectionScope() : this(default, default, default)
         {
         }
 
-        protected internal ClientConnectionScope(IServiceConnection outboundConnection, bool isDiagnosticClient)
+        protected internal ClientConnectionScope(HubServiceEndpoint endpoint, IServiceConnection outboundConnection, bool isDiagnosticClient)
         {
             // Only allow to carry one copy of connection properties regardless of how many nested scopes are created
             if (!IsScopeEstablished)
             {
                 _needCleanup = true;
+
+                // The lifetime of the async local we're about to create can be much longer than some of the objects we store inside.
+                // Instances of IServiceConnection can be stopped and replaced over time and nobody should keep references to the old ones.
+                // So we keep them inside async local wrapped in weak references to avoid unnecessarily prolonging their lifetime.
+                // Instances of HubServiceEndpoint (and connection container it references) are expected to exist for the duration
+                // of the process. So we can store them without wrapping in weak reference.
+                ConcurrentDictionary<HubServiceEndpoint, WeakReference<IServiceConnection>> dict = new ConcurrentDictionary<HubServiceEndpoint, WeakReference<IServiceConnection>>();
```

Comment:
In multiple endpoint hot reload cases, `HubServiceEndpoint` can be removed, shall we clean up the keys when the endpoint is removed? Or such little leak can be ignored?

Reply:
This indeed is a problem since we used HubServiceEndpoint as dictionary key (and store it unwrapped in async local). In case of hot reload the old HubServiceEndpoint  (which references all old connections) will continue flowing in the new execution contexts. The last PR fixes it by replacing HubServiceEndpoint with its unique id. As the result we won't leak HubServiceEndpoint. We still leak a dictionary entry (with long id and empty weakreference) per old endpoint but these don't carry any references to any heavy objects.